### ISSUE TITLE
Support MySQL inline INDEX specifications in DDLDatabase

### ISSUE THREAD

**Issue by bennofs** [jOOQ/jOOQ#9768]
### Expected behavior and actual behavior:

Expected: Codegen with DDLDatabase should translate the DDL to the correct H2 syntax.  
Actual behaviour: DDLDatabase creates an invalid H2 query. 

### Steps to reproduce the problem (if possible, create an MCVE: https://github.com/jOOQ/jOOQ-mcve):

Run Codegen with DDLDatabase on the file: 

```sql
CREATE TABLE flyway_test.author (
  id INT NOT NULL,
  first_name VARCHAR(50),
  last_name VARCHAR(50) NOT NULL,
  date_of_birth DATE,
  year_of_birth INT,
  address VARCHAR(50),

  INDEX year_of_birth (`year_of_birth`),
  CONSTRAINT pk_t_author PRIMARY KEY (ID)
);
```

This is a valid MySQL/MariaDB `CREATE TABLE` statement. However, running that with jooq-codegen and `DDLDatabase` produces:

```
Caused by: org.h2.jdbc.JdbcSQLNonTransientException: 
Unknown data type: "YEAR_OF_BIRTH"; 
SQL statement:
create table FLYWAY_TEST.AUTHOR(ID int not null, FIRST_NAME varchar(50) null, LAST_NAME varchar(50) not null, DATE_OF_BIRTH date null, YEAR_OF_BIRTH int null, ADDRESS varchar(50) null, constraint PK_T_AUTHOR primary key (ID), index YEAR_OF_BIRTH ("year_of_birth")) [50004-200]
```
So the index is not translated correctly. I modified the example to generate this error in https://github.com/bennofs/jOOQ/commit/cf3db8bb164d4ec34b59dbe4bdb9fc616d17d647. 

### Versions:

- jOOQ: master and 3.12.3
- Java: OpenJDK 64-Bit Server VM (build 11.0.6+10, mixed mode)
- Database (include vendor): MariaDB
- OS: Linux
- JDBC Driver (include name if inofficial driver): MySQL Connector/J

**Comment by jdimeo** [comment]
I'm experiencing the same problem I think because when DBeaver dumped my database, it included indexes as `KEY` clauses in the `CREATE TABLE` statement.

For example this:
```sql
CREATE TABLE `execution_dependencies` (
  `trigger_instance_id` varchar(64) NOT NULL,
  `dep_name` varchar(128) NOT NULL,
  `starttime` bigint(20) NOT NULL,
  `endtime` bigint(20) DEFAULT NULL,
  `dep_status` tinyint(4) NOT NULL,
  `cancelleation_cause` tinyint(4) NOT NULL,
  `project_id` int(11) NOT NULL,
  `project_version` int(11) NOT NULL,
  `flow_id` varchar(128) NOT NULL,
  `flow_version` int(11) NOT NULL,
  `flow_exec_id` int(11) NOT NULL,
  PRIMARY KEY (`trigger_instance_id`,`dep_name`),
  KEY `ex_end_time` (`endtime`)
) ENGINE=InnoDB DEFAULT CHARSET=utf8;
```
produces this error message:
[ERROR] Failed to execute goal org.jooq:jooq-codegen-maven:3.13.0:generate (default) on project intercept-rest: Error running jOOQ code generation tool: Error while exporting schema: SQL [create table "execution_dependencies"("trigger_instance_id" varchar(64) not null, "dep_name" varchar(128) not null, "starttime" bigint not null, "endtime" bigint null default null, "dep_status" tinyint not null, "cancelleation_cause" tinyint not null, "project_id" int not null, "project_version" int not null, "flow_id" varchar(128) not null, "flow_version" int not null, "flow_exec_id" int not null, primary key ("trigger_instance_id", "dep_name"), index "ex_end_time" ("endtime"))]; Unknown data type: "ex_end_time";
```

**Comment by lukaseder** [comment]
Thank you very much for your report, and sorry for the delay. This syntax currently translates to an anonymous block where it is supported, e.g. PostgreSQL (see https://www.jooq.org/translate):

```sql
do $$
begin
  create table FLYWAY_TEST.AUTHOR (
    ID int not null,
    FIRST_NAME varchar(50) null,
    LAST_NAME varchar(50) not null,
    DATE_OF_BIRTH date null,
    YEAR_OF_BIRTH int null,
    ADDRESS varchar(50) null,
    constraint PK_T_AUTHOR
      primary key (ID)
  );
  create index YEAR_OF_BIRTH on FLYWAY_TEST.AUTHOR("year_of_birth");
end $$
```

In H2, which is used behind the scenes by the `DDLDatabase` for now, there's no translation going on, which is why you're seeing the problem.

I'll think about whether this can be solved with the current `DDLDatabase` implementation and H2. The jOOQ Professional Edition can emulate anonymous blocks in H2 using the procedural API, but we're not using that yet, in these DDL emulations.

Ideally, we should interpret the DDL, but we're not there yet.

**Comment by cemo** [comment]
@lukaseder we are evaluating jOOQ for our needs. We hit a similar problem. What we would like to achieve is exact result of "show create table" statements of MySQL. Since key statement is included in the generated SQL, jOOQ raises error as in the example of @jdimeo. Any help please?

**Comment by lukaseder** [comment]
@cemo What kind of help are you looking for? As long as we can't handle this syntax, we can't handle this syntax :)

**Comment by cemo** [comment]
@lukaseder I am trying to understand now how it is working. As I understood, these DDL files are eventually stored as a in memory H2 Database. 

1. These SQL statements are translated to MySQL to H2? Or they must be compatible directly H2? 
2. I have used "show create table" on my MySQL DB to create necessary DDL statements. I used them as a source for this. Is this wrong approach?
3. What can kind of help we can provide to fix this issue? :) I can give a try with the help of your guidance :)

**Comment by lukaseder** [comment]
> I am trying to understand now how it is working. As I understood, these DDL files are eventually stored as a in memory H2 Database.

Yes, that's how it currently works. In the future, we'll interpret the DDL files ourselves, without an H2 database. Also, we'll offer spinning up a testcontainers database for code generation, which means we don't have to parse DDL ourselves: https://github.com/jOOQ/jOOQ/issues/6551

> 1. These SQL statements are translated to MySQL to H2? Or they must be compatible directly H2?

We translate them to H2. We currently don't have a query meta model representation for indexes being created along with tables. That's what this issue is for. It's a bigger task than just getting it to work for code generation. We need:

- A meta model (that already exists)
- A query meta model
- Being able to parse the MySQL syntax into the query meta model
- A way to interpret the query meta model
- A way to translate the query meta model

None of this is a problem, but all of it is work that hasn't been the top priority so far.

Especially because there are two very easy workarounds:

1. Use a separate, more standard `CREATE INDEX` command
2. Use the ignore tokens in your DDL, see https://www.jooq.org/doc/latest/manual/sql-building/dsl-context/custom-settings/settings-parser (could even be inserted using some simple regex-replacement)

> 2\. I have used "show create table" on my MySQL DB to create necessary DDL statements. I used them as a source for this. Is this wrong approach?

Not really wrong. We'll get there. But we aren't there yet.

> 3\. What can kind of help we can provide to fix this issue? :) I can give a try with the help of your guidance :)

Licensed customers can request reprioritisations. Other than that, I'd recommend the above workarounds. We generally don't accept PRs, especially not those that are as substantial as this one (see above "heavy" list of TODO bullets). The project management of ensuring a PR is correct is almost as big as just implementing it ourselves (and you don't have access to our tests anyway).

**Comment by cemo** [comment]
@lukaseder thank you for throughout explanation. I noticed some people are using TestContainers for replacing DDLDatabase way. I went in this way and successfully made a setup. It is working so good. :) I started to love more and more jOOQ 🤟

**Comment by lukaseder** [comment]
Sure, if that approach already works for you today (and it does if your database setup isn't super complex), then that's probably the most reliable way to go. No one parses vendor-specific SQL as well as the vendor :)

**Comment by lukaseder** [comment]
This is a similar problem as https://github.com/jOOQ/jOOQ/issues/12779, which talks about MySQL's inline column `COMMENT` syntax, which isn't supported by H2. #12779 is going to be solved for the commercial distributions using a new SQL transformation that decomposes DDL statements, see https://github.com/jOOQ/jOOQ/issues/13000 (Most recent SQL transformations are a commercial only feature, especially the ones transforming the new jOOQ 3.16 query object model)

This issue here will thus also be solved this way for now. Future versions of the jOOQ Open Source Edition may also support this feature once we can interpret the DDL ourselves, without translating it to H2.

**Comment by lukaseder** [comment]
The feature has now been implemented in jOOQ 3.17. It uses the commercial-only DDL query decomposition feature: https://github.com/jOOQ/jOOQ/issues/13000.

The jOOQ Open Source Edition will support this feature once we stop simulating DDL on H2 and interpret it, instead: https://github.com/jOOQ/jOOQ/issues/7034, which I'm hoping to implement soon, at least as an option.

Always a good reminder, too, to think about using testcontainers in your build to generate jOOQ code, instead of the `DDLDatabase`, as an alternative option: https://blog.jooq.org/using-testcontainers-to-generate-jooq-code/
